TVM Roadmap RFC (#50)
apache/tvm-rfcs: rfcs/0050-roadmaps.md

- Feature Name: roadmaps
- Start Date: 2022-01-07
- RFC PR: [apache/tvm-rfcs#0050](https://github.com/apache/tvm-rfcs/pull/0050)
- GitHub Issue: [apache/tvm#0000](https://github.com/apache/tvm/issues/0000)
- pre-RFC: https://discuss.tvm.apache.org/t/pre-rfc-tvm-roadmap/11171
- pre-RFC: https://discuss.tvm.apache.org/t/pre-rfc-roadmap-rfcs/11711
- Co-Authors: Denise Kutnick (@denise-k), Andrew Reusch (@areusch)

NOTE(areusch): This RFC is a combination of the two above pre-RFCs, and limited copy-editing has
been done to preserve the flow of the document.

## Summary

This RFC proposes to add product roadmaps to TVM.

Roadmaps should be seen as a way of unifying the planning process of TVM, all the way from ideas to

> PR merges. The roadmaps discussed in this pre-RFC are intentionally designed to integrate with TVM’s

Reviewer: ```suggestion
PR merges. The roadmaps discussed in this RFC are intentionally designed to integrate with TVM’s
```

Reviewer 2: done

> existing planning tools (e.g. Github tracking issues, RFCs), while adding an additional space for

Reviewer 2: done

sharing and collaboration earlier in the R&D process.

This proposal includes two categories of roadmaps:

> - **TVM Global Roadmap**: This roadmap aims to provide a holistic view of all components within TVM,

Reviewer: 1. Is the global roadmap any different than the sum of the component roadmaps?
2. How do we keep the global roadmap from falling into disrepair as teams update their own component roadmaps? a.k.a. who owns / drives it?

Reviewer 2: @driazati, to answer your questions:
1. The global roadmap is intended to be coarser-grained whereas the component roadmaps are intended to be finer-grained. If the tracking does ever overlap, both sets of roadmaps would be using the same base issues as roadmap items. For example, the TVM Global Roadmap and the TVM Automation Roadmap may both contain the [AutoTIR tracking issue](https://github.com/apache/tvm/issues/8473).
2. In general, upon introduction of any new roadmap, we are requesting that there is at least one maintainer who volunteers to keep the roadmap up to date. For the TVM Global Roadmap specifically, we anticipate that this roadmap will be maintained by several active contributors within the TVM Community, and will address this further in the TVM Global Roadmap RFC.

  focusing on features which broadly affect the entire project or add significant functionality.
- **TVM Component Roadmaps**: These roadmaps aim to provide a more detailed view of a subsection of
  TVM: including vertical efforts involving TVM's intermediate representations (such as Relay or
  TIR) and the horizontal efforts which span across TVM's full stack (such as Automation or
  Documentation). Details about each of the components in TVM will be discussed later in this

>   pre-RFC.

Reviewer: ```suggestion
  RFC.
```

Reviewer 2: done

## Motivations

- Integrate TVM’s technical vision and task tracking mechanisms together
- Provide everyone with visibility into all of TVM’s projects
- Enable individuals and organizations that contribute to TVM to influence the vision of TVM via
  roadmaps
- Encourage individuals and organizations that contribute to TVM to share the details of their
  longer-term efforts
- Ensure a openly-governed, equitable process towards creating new roadmaps for TVM

## Guide-Level Explanation

This guide-level explanation will describe the basic structure of the roadmap, without focusing
heavily on the intended contents.

### Tooling

The roadmap utilizes Github Projects as the underlying tooling mechanism. This maximizes reuse of

> existing content tracked in Github, and is very user-friendly for existing Github users.

Reviewer 2: done

    within the roadmap. Items on the roadmap generally have feature ownership at least partially
    identified at a minimum.
3. **Backlog Column**: Items on the roadmap which don't have an owner or a proposed development
   timeline fall into the **Backlog Items** column. This is also the default location where new
   roadmap issues are placed.

## Roadmap Process

This section proposes the **Roadmap-RFC** process, which aims to encapsulate the processes discussed
in the Overview section. The **Roadmap-RFC** is used to create, modify, or delete roadmaps. Details
on the process are listed below.

### Template of a Roadmap's Contextual Information

The following fields provide some contextual information for each roadmap.

- **Roadmap Name:** Name of the roadmap
- **Roadmap Maintainers**: These community members are primarily responsible for adding, modifying,
  and removing roadmap items. **Any community member can be listed as a maintainer of a roadmap,
  regardless of contributor status within the Apache organization.**
    - Note: If a Roadmap-RFC's proposed maintainer does not have TVM Committer or PMC status within

>     the Apache organization, they will receive an invitation for a Triage role in Apache TVM as soon

Reviewer: Adding labels is a strong tool for organizing issues as they go into roadmap items, I think everyone that is a roadmap maintainer should have the triage role if nothing else, regardless of who else happens to be on the roadmap

    the roadmap.
- For each theme, include a set of definitions specific to the proposed roadmap.
    - What does this theme mean in the context of this roadmap?
    - Are there multiple definitions for this theme? For example, `performance` could be interpreted
    as tuning times, runtime latency, or a number of other definitions.
- For each theme, include a set of success criteria specific to the proposed roadmap.
    - What types of metrics would be relevant to this roadmap?

### Adding Items to a Roadmap

A roadmap's maintainers are the primary folks responsible for adding, modifying, and removing items
for a roadmap. This isn't a hard and fast rule—for example, it may be expedient for other community
members to triage new roadmap items into a roadmap's backlog. However, the maintainers should be
considered the "owners" of a roadmap, and generally no rigid process is defined around modifying the
items in a roadmap.

Roadmaps are defined using [GitHub
Projects](https://docs.github.com/en/issues/trying-out-the-new-projects-experience/about-projects)
and can include GitHub Issues, Pull Requests, and simple note cards. Maintainers should strive to

> place mainly **GitHub Issues** in Roadmaps to make it possible for the community to learn more about

Reviewer 3: ```suggestion
place mainly **GitHub Issues** in roadmaps to make it possible for the community to learn more about
```

Reviewer 2: fixed!

ongoing work and reduce triage burden.

Each item on a roadmap is intended to track one of these community processes:

- **pre-RFCs** serve as a way to begin discussions on planned work at the earliest stage of
  maturity. **pre-RFCs** are typically posted in the [TVM Discuss
  forums](http://discuss.tvm.apache.org) in order to solicit TVM Community feedback. For an example

>   of a **pre-RFC**, see the screenshot of Andrew R's proposal to *Convert RST Docs to Markdown*

Reviewer: ```suggestion
  of a **pre-RFC**, see the screenshot of @areusch's proposal to [Convert RST Docs to Markdown](https://discuss.tvm.apache.org/t/docs-discuss-convert-restructuredtext-docs-to-markdown/10264)
```

Reviewer 2: done, thanks!

>   of a **pre-RFC**, see the screenshot of Andrew R's proposal to *Convert RST Docs to Markdown*

Reviewer 3: Are we expecting pre-RFCs to be this light weight? Following the same template as the eventual RFC both ensures people are considering the areas in the RFC process as well as starting the process of creating that eventual RFC document.

Reviewer 2: This came up in some of our previous discussions, and we've purposefully left this ambiguous because pre-RFC structures/guidelines are outside the scope of this RFC. 

I do 💯 agree that there should be a set of guidelines for pre-RFCs, and I'd love to discuss this further. Perhaps it can be a topic in the next TVM community meeting! 😸

  below.

    **pre-RFCs** can be tracked on a Roadmap by preemptively creating a GitHub Task-tracking Issue

>     in [tvm-rfcs](https://github.com/apache/tvm-rfcs).

Reviewer 3: Does it not need to be in `apache/tvm` to enter the roadmap?

Reviewer 2: No, it doesn't need to be. The reason for this is because (for permissions reasons) the roadmaps will actually be created at the [apache top level](https://github.com/orgs/apache/projects) which can link to any repository within apache. Illustration below:

![image](https://user-images.githubusercontent.com/50588693/149224231-2da1655b-77c8-4c37-a9f5-5c6b8bb802bf.png)

    ![image|690x460](./assets/0050/roadmap-item-pre-rfc.png)

- **RFCs** serve as a way to share feature proposals with the TVM Community. The
  [tvm-rfcs](https://github.com/apache/tvm-rfcs) repo is used for the creation and acceptance of
  **RFCs**. For some examples of **RFCs**, see the screenshot of open pull requests in
  [tvm-rfcs](https://github.com/apache/tvm-rfcs) below.

    Open **RFCs** can be directly linked into any roadmap. Once an **RFC** is accepted, please use
    the **Github Task-Tracking** process to track **RFC** Execution.
    ![image|690x357](./assets/0050/roadmap-item-rfc.png)

- **Github Task-Tracking Issues** are used in [tvm](https://github.com/apache/tvm) to share the
  progress of accepted RFCs over time. For an example of a **Github Task-Tracking Issue**, see the

>   screenshot of Andrew L's RFC to *Add Mixed-Precision Support to TVM* below.

Reviewer 3: Would be good use the GitHub username here to illustrate which Andrew L further :smile_cat:

Reviewer 2: :%s/Andrew L's/@AndrewZhaoLuo's/g

Reviewer 2: done!

    These task-tracking issues can be directly linked into any roadmap.
    ![image|663x499](./assets/0050/roadmap-item-gh-task-tracking-issue.png)

- **Bugfixes** are actionable GitHub Issues not necessarily connected to a Task-Tracking
  Issue. Generally, they require only 1 PR and the work is clearly specified in the issue. For an
  example of a **bugfix** Github Issue, see the screenshot of a flaky CI test report below.

    You can directly link **bugfixes** in a roadmap by adding their corresponding GitHub Issues as a
    card. Generally, it’s the maintainer’s responsibility to combine related **bugfixes** in order
    to keep the roadmap concise. For example, if you have 10 flaky tests in CI, it might make sense
    to have a global tracking issue for flaky CI tests which points to each individual bug.
     ![image|690x368](./assets/0050/roadmap-item-bugfix.png)

> 

Reviewer: There is a category of mid-sized issues missing here that people generally don't file right now, but there is a lot of work not directly related to implementing an RFC or a bug fix that goes on (i.e. bigger than a bug fix but smaller than an RFC). I don't think there's any harm in making these into issues, even if they're not actionable/owned right away or if the ideas are less well formed (maybe the issue is just so a specific developer doesn't forget about something).

This would be nice with a wider application of the triage role since devs writing issues could assign it to themselves or mark it as triaged so it doesn't fill the queue.

Reviewer 2: This is a great callout. I agree that a roadmap's maintainers should have the ability to track these types of issues on the roadmap.

Perhaps this falls under Github Task-Tracking, and we should just change the wording so that we're not boxing that particular category in as "accepted RFCs only" - what are your thoughts @driazati @areusch?

Reviewer 3: I think that'd be good @denise-k, we've done this in the past to track future work arising from a PR review (for example: https://github.com/apache/tvm/issues/8792) - this should serve as a reminder and incentive to follow up :smile_cat:

Reviewer 2: Thanks! I've added a mini-section here to address this.

  functionality than Github projects. However, this would require much more effort to create and
  maintain.

## Prior Art

This RFC is inspired by the previous work done to improve the tracking mechanisms of TVM. This
includes but is not limited to the existing RFC process, release process, and technical visions of
TVM.

![image|690x351](./assets/0050/prior-art-1.jpeg)

It is also inspired by the [public Github roadmap](https://github.com/github/roadmap), which is an
excellent example of usage for Github Projects.

![image|690x426](./assets/0050/prior-art-public-gh-roadmap.png)

## Upstreaming Timeline

This is a prospective upstreaming timeline for the TVM roadmap.

|Milestone Name|Deliverables|Schedule|
| --- | --- | --- |
|TVM Roadmap M1A: Taking the Plunge|<li>All roadmap projects defined<br><li>Initial prototype roadmap (TVM CI & Testing) upstreamed <br> <li>Skeleton roadmap documentation upstreamed <br> <li>Scope of M1B and M1C clearly defined|Late September|
|TVM Roadmap M1B: Building Momentum|<li> At least 3 additional roadmaps upstreamed <br> <li> Additional roadmap process documentation upstreamed|Mid October|

> |TVM Roadmap M1C: Ready for Prime Time|<li> All initial roadmaps (described in M1A RFC) upstreamed <br> <li> All RFC opens from M1A and M1B addressed <br><li> Publicize roadmap in documentation and TVM forums|Early November|

Reviewer: These dates seem off

Reviewer 2: @areusch was going to remove the dates from this

Reviewer 2: fixed

## Unresolved Questions

Questions about the design and contents of the TVM Roadmap will be progressively resolved through
the upstreaming timeline shown above. Any unresolved questions at the end of TVM Roadmap Milestone 1

> will be created as roadmap items in the TVM Community Roadmap 🙂

Reviewer 3: The last character here doesn't render for me.

Reviewer 2: It's a smile emoji; let's go ahead and remove it to avoid any rendering issues.

Reviewer 2: done!

## Future Possibilities

- Create [Github issue templates](https://github.com/apache/tvm/tree/main/.github/ISSUE_TEMPLATE)
  for roadmap items (similar to the existing Github issue templates for docs, CI, etc).
- Use [the new Github issues/projects (closed beta)](https://github.com/features/issues) to improve
  roadmap functionality.

## Special Thanks

> Thanks so much for copyediting, @areusch and @electriclilies! :hugs:

Reviewer 3: Should this be @denise-k and @electriclilies given @areusch is the author here? :smile_cat:

Reviewer 2: thanks for bringing this up - I authored this initially (which is why I didn't refer to myself in third person), and @areusch ported it over to this RFC + made some additions. @areusch and I are already both listed as coauthors, so really it should be just @electriclilies listed as a copyeditor.
